### PR TITLE
RSA max strength fixes

### PR DESCRIPTION
**Pull request by paulidale**
- Remove the limit on the maximum strength of RSA keys
- Add a check during RSA key generation to not allow a key to be generated that has a higher security strength than the random number generator being used
- Put back some tests that got lost

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by slontis** [comment]
Your source changes look fine.. Will wait and see if the tests pass :)

**Comment by paulidale** [comment]
Tests have passed after a bit of fiddling around deprecations.

File: test/recipes/15-test_genrsa.t [hunk 1]
```diff
@@ -105,18 +102,14 @@ ok(!run(app([ 'openssl', 'genpkey', '-propquery', 'unknown',
    "genpkey requesting unknown=yes property should fail");
 
 
- SKIP: {
-    skip "Skipping rsa command line test", 4 if disabled("deprecated-3.0");
-
-    ok(run(app([ 'openssl', 'genrsa', '-3', '-out', 'genrsatest.pem', $good ])),
```
Comment (slontis) [inline test/recipes/15-test_genrsa.t:111]: Err.. the old tests were testing the -3 option.. Now nothing probably does this.

Comment (paulidale) [reply test/recipes/15-test_genrsa.t:111]: Scroll up :)  This is testing a minimal sized key with `-3`.

File: test/recipes/15-test_genrsa.t [hunk 2]
```diff
@@ -25,18 +25,21 @@ my $no_fips = disabled('fips') || ($ENV{NO_FIPS} // 0);
...
-                 '-pkeyopt', 'rsa_keygen_pubexp:3'])),
-               0, "genrsa -3 8");
-} else {
-    is(run(app([ 'openssl', 'genrsa', '-3', '-out', 'genrsatest.pem', '8'])),
-               0, "genrsa -3 8");
-}
+is(run(app([ 'openssl', 'genpkey', '-out', 'genrsatest.pem',
+             '-algorithm', 'RSA', '-pkeyopt', 'rsa_keygen_bits:8',
+             '-pkeyopt', 'rsa_keygen_pubexp:3'])),
+           0, "genpkey -3 8");
```
Comment (slontis) [inline test/recipes/15-test_genrsa.t:34]: Update the description of the tests... -3 is not right, and each test should have a different name

File: test/recipes/15-test_genrsa.t (hunk 1 above)
Comment (paulidale) [reply test/recipes/15-test_genrsa.t:111]: Put this back for deprecated builds only.

File: test/recipes/15-test_genrsa.t (hunk 2 above)
Comment (paulidale) [reply test/recipes/15-test_genrsa.t]: This was preexisting :(
Fixed.

**Review by slontis — APPROVED** [review]
Approved assuming tests pass.

File: crypto/err/openssl.txt [hunk 3]
```diff
@@ -1190,6 +1190,8 @@ RSA_R_PSS_SALTLEN_TOO_SMALL:164:pss saltlen too small
 RSA_R_PUB_EXPONENT_OUT_OF_RANGE:178:pub exponent out of range
 RSA_R_P_NOT_PRIME:128:p not prime
 RSA_R_Q_NOT_PRIME:129:q not prime
+RSA_R_RANDOMNESS_SOURCE_STRENGTH_INSUFFICIENT_FOR_KEYGEN:180:\
```
Comment (t8m) [inline crypto/err/openssl.txt:1193]: Now that is a looong reason code name. What about just dropping the `_FOR_KEYGEN` part?

**Comment by t8m** [comment]
Also needs rebase :(

File: crypto/err/openssl.txt (hunk 3 above)
Comment (paulidale) [reply crypto/err/openssl.txt]: Shortened this.

File: include/openssl/rsaerr.h [hunk 4]
```diff
@@ -85,6 +85,7 @@
 # define RSA_R_PUB_EXPONENT_OUT_OF_RANGE                  178
 # define RSA_R_P_NOT_PRIME                                128
 # define RSA_R_Q_NOT_PRIME                                129
+# define RSA_R_RANDOMNESS_SOURCE_STRENGTH_INSUFFICIENT 180
```
Comment (t8m) [inline include/openssl/rsaerr.h:88]: Nit: Alignment?

File: crypto/rsa/rsa_err.c [hunk 5]
```diff
@@ -118,6 +118,8 @@ static const ERR_STRING_DATA RSA_str_reasons[] = {
     "pub exponent out of range"},
     {ERR_PACK(ERR_LIB_RSA, 0, RSA_R_P_NOT_PRIME), "p not prime"},
     {ERR_PACK(ERR_LIB_RSA, 0, RSA_R_Q_NOT_PRIME), "q not prime"},
+    {ERR_PACK(ERR_LIB_RSA, 0, RSA_R_RANDOMNESS_SOURCE_STRENGTH_INSUFFICIENT),
+    "randomness source strength insufficient for keygen"},
```
Comment (t8m) [inline crypto/rsa/rsa_err.c:122]: not sure if the `for keygen` should be also dropped from the text

File: include/openssl/rsaerr.h (hunk 4 above)
Comment (paulidale) [reply include/openssl/rsaerr.h]: Fixed.

File: crypto/rsa/rsa_err.c (hunk 5 above)
Comment (paulidale) [reply crypto/rsa/rsa_err.c]: Dropped

File: crypto/err/openssl.txt [hunk 6]
```diff
@@ -1190,6 +1190,8 @@ RSA_R_PSS_SALTLEN_TOO_SMALL:164:pss saltlen too small
 RSA_R_PUB_EXPONENT_OUT_OF_RANGE:178:pub exponent out of range
 RSA_R_P_NOT_PRIME:128:p not prime
 RSA_R_Q_NOT_PRIME:129:q not prime
+RSA_R_RANDOMNESS_SOURCE_STRENGTH_INSUFFICIENT:180:\
+	randomness source strength insufficient for keygen
```
Comment (t8m) [inline crypto/err/openssl.txt:1194]: `for keygen` still left here

Comment (paulidale) [reply crypto/err/openssl.txt]: Sigh....
Fixed now.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
@slontis still OK?

**Comment by slontis** [comment]
Reapproved.

**Comment by paulidale** [comment]
Merged to master.

This is addressing a security concern: generating a key stronger than 256 bits simply isn't possible using our current RNG setup.

**Comment by slontis** [comment]
> This is addressing a security concern: generating a key stronger than 256 bits simply isn't possible using our current RNG setup.

At some point we need a rand_get_bytes_ex() like method that can pass a minimum bit strength (along with code from 90B that can generate more than 256 bits)

**Pull request closed without merging**